dashboard/signals.py

```diff
@@ -25,27 +38,61 @@ def precreate_programenrollment(sender, instance, **kwargs):  # pylint: disable=
         instance_in_db = CachedEnrollment.objects.filter(id=instance.id).exclude(data__isnull=True).count()
         # if the count is 1, it means the student unenrolled from the course run
         if instance_in_db == 1:
-            active_enrollment_count = CachedEnrollment.objects.filter(
-                user=user,
-                course_run__course__program=program
-            ).exclude(data__isnull=True).count()
             # if there is only one enrollment with data non None, it means that it is the
             # current instance is the only one for the program, so the program enrollment
             # needs to be deleted
-            if active_enrollment_count <= 1:  # theoretically this cannot be <1, but just in case
+            if CachedEnrollment.active_count(user, program) <= 1:  # theoretically this cannot be <1, but just in case
                 ProgramEnrollment.objects.filter(
                     user=user,
                     program=program
                 ).delete()
 
 
-@receiver(post_save, sender=CachedEnrollment, dispatch_uid="update_programenrollment")
-def create_programenrollment(sender, instance, **kwargs):  # pylint: disable=unused-argument
+@receiver(post_save, sender=CachedEnrollment, dispatch_uid="cachedenrollment_post_save")
+def handle_update_enrollment(sender, instance, **kwargs):  # pylint: disable=unused-argument
     """
-    Signal handler to create Program enrollment when the CachedEnrollment table is updated
+    Create ProgramEnrollment when a CachedEnrollment is created/updated, and update the index
     """
     if instance.data is not None:
-        ProgramEnrollment.objects.get_or_create(
+        program_enrollment = ProgramEnrollment.objects.get_or_create(
             user=instance.user,
             program=instance.course_run.course.program
-        )
+        )[0]
+        index_program_enrolled_users.delay([program_enrollment])
+
+
+@receiver(post_save, sender=CachedCertificate, dispatch_uid="cachedcertificate_post_save")
+def handle_update_certificate(sender, instance, **kwargs):  # pylint: disable=unused-argument
+    """When a CachedCertificate model is updated, update index """
+    if instance.data is not None:
+        program_enrollment = ProgramEnrollment.objects.get_or_create(
+            user=instance.user,
+            program=instance.course_run.course.program
+        )[0]
+        index_program_enrolled_users.delay([program_enrollment])
+
+
+@receiver(post_delete, sender=CachedEnrollment, dispatch_uid="cachedenrollment_post_delete")
+def handle_delete_enrollment(sender, instance, **kwargs):  # pylint: disable=unused-argument
+    """Update index when CachedEnrollment model instance is deleted."""
+    user = instance.user
+    program = instance.course_run.course.program
+    program_enrollment = ProgramEnrollment.objects.filter(user=user, program=program).first()
+    if program_enrollment and CachedEnrollment.active_count(user, program) == 0:
```

Comment:
if the code `if program_enrollment` means `if program_enrollment is not None` you should be explicit